source/adapters/opencl/context.hpp

```diff
@@ -10,9 +10,52 @@
 #pragma once
 
 #include "common.hpp"
+#include "device.hpp"
 
-namespace cl_adapter {
-ur_result_t
-getDevicesFromContext(ur_context_handle_t hContext,
-                      std::unique_ptr<std::vector<cl_device_id>> &DevicesInCtx);
-}
+#include <vector>
+
+struct ur_context_handle_t_ {
+  using native_type = cl_context;
+  native_type Context;
+  std::vector<ur_device_handle_t> Devices;
+  uint32_t DeviceCount;
+
+  ur_context_handle_t_(native_type Ctx, uint32_t DevCount,
+                       const ur_device_handle_t *phDevices)
+      : Context(Ctx), DeviceCount(DevCount) {
+    for (uint32_t i = 0; i < DeviceCount; i++) {
+      Devices.emplace_back(phDevices[i]);
+    }
+  }
+
+  static ur_result_t makeWithNative(native_type Ctx, uint32_t DevCount,
+                                    const ur_device_handle_t *phDevices,
+                                    ur_context_handle_t &Context) {
+    auto URContext =
+        std::make_unique<ur_context_handle_t_>(Ctx, DevCount, phDevices);
+    native_type &NativeContext = URContext->Context;
+    uint32_t &DeviceCount = URContext->DeviceCount;
+    if (!DeviceCount) {
+      CL_RETURN_ON_FAILURE(
+          clGetContextInfo(NativeContext, CL_CONTEXT_NUM_DEVICES,
+                           sizeof(DeviceCount), &DeviceCount, nullptr));
+      std::vector<cl_device_id> CLDevices(DeviceCount);
+      CL_RETURN_ON_FAILURE(clGetContextInfo(NativeContext, CL_CONTEXT_DEVICES,
+                                            sizeof(CLDevices), CLDevices.data(),
+                                            nullptr));
+      URContext->Devices.resize(DeviceCount);
+      for (uint32_t i = 0; i < DeviceCount; i++) {
+        ur_native_handle_t NativeDevice =
+            reinterpret_cast<ur_native_handle_t>(CLDevices[i]);
+        UR_RETURN_ON_FAILURE(urDeviceCreateWithNativeHandle(
+            NativeDevice, nullptr, nullptr, &(URContext->Devices[i])));
+      }
+    }
+    Context = URContext.release();
+    return UR_RESULT_SUCCESS;
+  }
+
+  ~ur_context_handle_t_() {}
```

Comment:
Should this be calling `clReleaseContext`?

Reply:
Ah, I think I misunderstood that, that's a good point, I didn't link that with reference counting, it actually would make sense to call `clReleaseContext` with the destructor as the ur retaining will decrease the reference counting and delete the resources.